### PR TITLE
Add GLM4_MOE model support

### PR DESCRIPTION
**Pull request by vvvdwbvvv**
## Summary

This PR adds support for GLM4.5 (GLM-4 MOE) models to the Liger Kernel #951 
https://huggingface.co/zai-org/GLM-4.5 which share the same structure as GLM 4.6

## Testing Done

For the convergence test on fp32, model size can easily leads to OOM, initially I was using 4090 to run the tests, however only fp32 encounters OOM, so I move forward to L40S to finish all the tests. 

- Hardware Type: <L40S>
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [ ] run `make test-convergence` to ensure convergence

### CODE REVIEW

File: src/liger_kernel/transformers/model/glm4_moe.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+    if skip_logits and labels is None and shift_labels is None:
+        raise ValueError("skip_logits is True, but labels and shift_labels are None")
+
+    if skip_logits is None:
+        # By default, if in training mode, don't materialize logits
+        skip_logits = self.training and (labels is not None or shift_labels is not None)
+
+    if skip_logits:
+        loss = LigerForCausalLMLoss(
```
Comment (kashif) [inline src/liger_kernel/transformers/model/glm4_moe.py]: kindly have a look at the other model examples and adapt to new API that returns the metric

**Comment by vvvdwbvvv** [comment]
Fixed in [5af9d16](https://github.com/linkedin/Liger-Kernel/pull/952/commits/5af9d164b171cc1a6224217d8ebfd13a6c267cc9)

File: src/liger_kernel/transformers/monkey_patch.py [hunk 2]
```diff
@@ -2065,6 +2065,81 @@ def apply_liger_kernel_to_glm4(
...
+            # patch MOE layers
+            if isinstance(decoder_layer.mlp, Glm4MoeMoE):
+                experts = decoder_layer.mlp.experts
+                if experts is not None:
+                    for expert in experts:
+                        _patch_swiglu_module(expert, LigerSwiGLUMLP)
+
+                shared_experts = decoder_layer.mlp.shared_experts
+                if shared_experts is not None:
+                    _patch_swiglu_module(shared_experts, LigerSwiGLUMLP)
```
Comment (Tcc0403) [inline src/liger_kernel/transformers/monkey_patch.py]:
```suggestion
            if swiglu:
	            if isinstance(decoder_layer.mlp, Glm4MoeMoE):
	                experts = decoder_layer.mlp.experts
	                if experts is not None:
	                    for expert in experts:
	                        _patch_swiglu_module(expert, LigerSwiGLUMLP)
	
	                shared_experts = decoder_layer.mlp.shared_experts
	                if shared_experts is not None:
	                    _patch_swiglu_module(shared_experts, LigerSwiGLUMLP)
	             elif isinstance(decoder_layer.mlp, Glm4MoeMLP):
                	_patch_swiglu_module(decoder_layer.mlp, LigerSwiGLUMLP)
            if rms_norm:
                _patch_rms_norm_module(decoder_layer.input_layernorm)
                _patch_rms_norm_module(decoder_layer.post_attention_layernorm)
```
You need to check whether this layer is MLP or MoE before patching, plus MoE patching should be under `if swiglu` scope.

File: test/transformers/test_monkey_patch.py [hunk 3]
```diff
@@ -2579,6 +2589,79 @@ def test_apply_liger_kernel_to_instance_for_glm4():
...
+            head_dim=8,
+            first_k_dense_replace=1,
+        )
+        dummy_model_instance = AutoModelForCausalLM.from_config(config)
+
+        # Check that model instance variables are not yet patched with Liger modules
+        assert inspect.getsource(dummy_model_instance.forward) != inspect.getsource(glm4_moe_lce_forward)
+        assert inspect.getsource(dummy_model_instance.model.norm.forward) != inspect.getsource(LigerRMSNorm.forward)
+        for decoder_layer in dummy_model_instance.model.layers:
+            assert inspect.getsource(decoder_layer.mlp.forward) != inspect.getsource(LigerSwiGLUMLP.forward)
```
Comment (Tcc0403) [inline test/transformers/test_monkey_patch.py]: Same as above, `mlp.forward` shouldn't be patched if this is a moe layer. Need to check `isinstance()` first.

File: test/transformers/test_monkey_patch.py [hunk 4]
```diff
@@ -2579,6 +2589,79 @@ def test_apply_liger_kernel_to_instance_for_glm4():
...
+                    )
+
+        # Test applying kernels to the model instance
+        _apply_liger_kernel_to_instance(model=dummy_model_instance)
+
+        # Check that the model's instance variables were correctly patched with Liger modules
+        assert inspect.getsource(dummy_model_instance.forward) == inspect.getsource(glm4_moe_lce_forward)
+        assert inspect.getsource(dummy_model_instance.model.norm.forward) == inspect.getsource(LigerRMSNorm.forward)
+        for decoder_layer in dummy_model_instance.base_model.layers:
+            if decoder_layer.mlp is not None:
```
Comment (Tcc0403) [inline test/transformers/test_monkey_patch.py]: check `isinstance(decoder_layer.mlp, GLM4MoeMLP)` instead of just `is not None`.

File: test/transformers/test_monkey_patch.py [hunk 5]
```diff
@@ -2579,6 +2589,79 @@ def test_apply_liger_kernel_to_instance_for_glm4():
...
+                assert inspect.getsource(decoder_layer.mlp.forward) == inspect.getsource(LigerSwiGLUMLP.forward)
+                assert inspect.getsource(decoder_layer.post_attention_layernorm.forward) == inspect.getsource(
+                    LigerRMSNormForGlm4.forward
+                )
+                assert inspect.getsource(decoder_layer.input_layernorm.forward) == inspect.getsource(
+                    LigerRMSNormForGlm4.forward
+                )
```
Comment (Tcc0403) [inline test/transformers/test_monkey_patch.py]: Wrong scope, rms norm should be checked in both mlp layer and moe layer

**Comment by vvvdwbvvv** [comment]
@Tcc0403 Fixed in 876d075

File: src/liger_kernel/transformers/monkey_patch.py [hunk 6]
```diff
@@ -2140,6 +2140,91 @@ def apply_liger_kernel_to_glm4(
...
```
Comment (Tcc0403) [inline src/liger_kernel/transformers/monkey_patch.py]: It's not under `if swiglu:` scope

Comment (vvvdwbvvv) [reply src/liger_kernel/transformers/monkey_patch.py]: Thank you, I'll fix it

**Comment by vvvdwbvvv** [comment]
@Tcc0403 Fixed in [cf9f212](https://github.com/linkedin/Liger-Kernel/pull/952/commits/cf9f21243f8c42b8581e19b0474efe51f335f03b)

File: test/convergence/bf16/test_mini_models.py [hunk 7]
```diff
@@ -1128,6 +1138,37 @@
...
+            num_key_value_heads=2,
+            rms_norm_eps=1e-5,
+            rope_scaling=None,
+            rope_theta=500_000,
+            tie_word_embeddings=False,
+            use_cache=True,
+            vocab_size=32000,  # 151552
+            attention_bias=True,
+            attn_implementation="sdpa",  # default value, pytorch native attention
+        ),
```
Comment (Tcc0403) [inline test/convergence/bf16/test_mini_models.py:1177]: Set smaller experts related numbers as well
```
"moe_intermediate_size": 1408,
"num_experts_per_tok": 2,
"n_shared_experts": 1,
"n_routed_experts": 8,
"routed_scaling_factor": 1.0,
"n_group": 1,
"topk_group": 1,
"first_k_dense_replace": 1,
```

File: test/convergence/bf16/test_mini_models_with_logits.py [hunk 8]
```diff
@@ -1075,6 +1085,38 @@
             attn_implementation="sdpa",  # default value, pytorch native attention
         ),
     )
+
+if GLM4_MOE_AVAILABLE:
+    MINI_MODEL_SETUPS["mini_glm4_moe"] = MiniModelConfig(
+        liger_kernel_patch_func=apply_liger_kernel_to_glm4_moe,
+        liger_kernel_patch_revert_func=revert_liger_kernel_to_glm4_moe,
+        model_class=Glm4MoeForCausalLM,
+        mini_model_config=Glm4MoeConfig(
```
Comment (Tcc0403) [inline test/convergence/bf16/test_mini_models_with_logits.py:1094]: ditto

File: test/convergence/fp32/test_mini_models.py [hunk 9]
```diff
@@ -1064,6 +1074,37 @@
         ),
     )
 
+if GLM4_MOE_AVAILABLE:
+    MINI_MODEL_SETUPS["mini_glm4_moe"] = MiniModelConfig(
+        liger_kernel_patch_func=apply_liger_kernel_to_glm4_moe,
+        liger_kernel_patch_revert_func=revert_liger_kernel_to_glm4_moe,
+        model_class=Glm4MoeForCausalLM,
+        mini_model_config=Glm4MoeConfig(
```
Comment (Tcc0403) [inline test/convergence/fp32/test_mini_models.py:1082]: ditto

File: test/convergence/fp32/test_mini_models_with_logits.py [hunk 10]
```diff
@@ -1093,6 +1103,37 @@
         ),
     )
 
+if GLM4_MOE_AVAILABLE:
+    MINI_MODEL_SETUPS["mini_glm4_moe"] = MiniModelConfig(
+        liger_kernel_patch_func=apply_liger_kernel_to_glm4_moe,
+        liger_kernel_patch_revert_func=revert_liger_kernel_to_glm4_moe,
+        model_class=Glm4MoeForCausalLM,
+        mini_model_config=Glm4MoeConfig(
```
Comment (Tcc0403) [inline test/convergence/fp32/test_mini_models_with_logits.py:1111]: ditto

**Review by Tcc0403 — CHANGES_REQUESTED** [review]
Patching should work, but we need to tune down moe related numbers to speed up convergence tests.

**Comment by vvvdwbvvv** [comment]
@Tcc0403  Fixed at 375903c, thank you.

**Review by Tcc0403 — COMMENTED** [review]
Although the current mini model configuration can verify the correctness of the patching, the training loss remains unchanged (>10) across all iterations (the loss curve does not decrease). I suggest commenting out this test case for now, adding a TODO note, and opening an issue to track and investigate this problem.
